### DISCUSSION TITLE
Ideas: Pluggable learning kernel strategy — STDP, BCM, Oja, and EWC

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#234] [Ideas]
## Ideas: Pluggable Learning Kernel Strategy — STDP, BCM, Oja, and EWC

Issue #181 and Discussion #227 introduced the `STDPOnlineLearner` with a fixed STDP timing kernel. Discussion #231 wired it into the `LEARNING` phase of `CognitiveCycle`. Now there's an open question raised in #227:

> Should the STDP kernel itself be pluggable? What about BCM rule, Oja's rule, or EWC (Elastic Weight Consolidation)?

This thread collects design options for making ASI:BUILD's online learning kernel swappable.

---

### Background: What's Fixed vs. What's Pluggable Today

The current `STDPOnlineLearner` (from #181 + wiki Phase-5-Online-Learning.md):

```python
class OnlineLearnerBase(ABC):
    @abstractmethod
    async def on_weight_update(self, update: WeightUpdate) -> None: ...
    @abstractmethod
    async def tick(self, bb: Blackboard) -> None: ...

class STDPOnlineLearner(OnlineLearnerBase):
    A_PLUS  = 0.01   # LTP amplitude
    A_MINUS = 0.012  # LTD amplitude
    TAU_PLUS  = 20.0 # ms — LTP time constant
    TAU_MINUS = 20.0 # ms — LTD time constant
```

The STDP kernel is hardcoded. The learning rule is not injectable.

---

### Option A — Kernel Protocol / Strategy Pattern

Introduce a `LearningKernel` protocol with a single method:

```python
from typing import Protocol

class LearningKernel(Protocol):
    def compute_delta(self, pre_t: float, post_t: float, weight: float) -> float:
        """Return the weight delta given spike timing and current weight."""
        ...
```

Concrete kernels:

| Kernel | Rule | Key params |
|--------|------|-----------|
| `STDPKernel` | Spike-timing dependent plasticity | `A+`, `A-`, `τ+`, `τ-` |
| `BCMKernel` | Bienenstock-Cooper-Munro | `theta_M` (sliding threshold) |
| `OjaKernel` | Oja's rule — normalised Hebbian | `eta` (learning rate) |
| `EWCKernel` | Elastic Weight Consolidation | `lambda`, `Fisher` matrix diagonal |

`OnlineLearnerBase` accepts a `kernel: LearningKernel` in its constructor:

```python
class OnlineLearnerBase(ABC):
    def __init__(self, kernel: LearningKernel | None = None) -> None:
        self.kernel = kernel or STDPKernel()  # default unchanged
```

**Pros**: Zero-cost when using default. Fully testable by injecting a mock kernel.  
**Cons**: Adds one layer of indirection. BCM and EWC have different signatures (BCM needs θ_M history, EWC needs Fisher).

---

### Option B — Registry + Config-Driven Selection

Define a `LearningKernelRegistry` keyed on string names:

```python
KERNEL_REGISTRY: dict[str, type[LearningKernel]] = {
    "stdp":  STDPKernel,
    "bcm":   BCMKernel,
    "oja":   OjaKernel,
    "ewc":   EWCKernel,
}

def build_online_learner(config: dict) -> OnlineLearnerBase:
    kernel_cls = KERNEL_REGISTRY[config.get("kernel", "stdp")]
    kernel = kernel_cls(**config.get("kernel_params", {}))
    return STDPOnlineLearner(kernel=kernel)  # or a generic learner
```

Driven by `asi-build.yaml`:

```yaml
online_learning:
  kernel: ewc
  kernel_params:
    lambda: 400
    fisher_update_interval: 100  # ticks
```

**Pros**: Operator-selectable at runtime without code changes. Easy A/B testing.  
**Cons**: EWC requires Fisher matrix maintenance — significantly more state. Config coupling.

---

### Option C — Composite Kernel (Ensemble)

Allow stacking kernels with weighted contributions:

```python
class CompositeKernel:
    def __init__(self, kernels: list[tuple[LearningKernel, float]]) -> None:
        self.kernels = kernels

    def compute_delta(self, pre_t, post_t, weight):
        return sum(w * k.compute_delta(pre_t, post_t, weight)
                   for k, w in self.kernels)
```

Example: 70% STDP + 30% Oja's rule for a normalised-but-temporally-sensitive learner.

**Pros**: Maximum flexibility. Models real cortical plasticity better.  
**Cons**: Hard to debug. Interactions between kernels are non-obvious. Probably over-engineering for Phase 5.

---

### EWC: Special Consideration

EWC (Kirkpatrick et al., 2017) is different from STDP/BCM/Oja because it is a **catastrophic forgetting prevention** mechanism, not just a plasticity rule:

```
L_EWC = L_new + (λ/2) * Σ_i F_i (θ_i - θ*_i)²
```

Where `F_i` is the diagonal of the Fisher information matrix for parameter `i`, and `θ*_i` is the previous task's optimal parameter.

For ASI:BUILD, EWC would protect the `INTEGRATION` phase's consolidated KG embeddings from being overwritten by new `LEARNING` phase updates. This maps naturally to the `SLEEP_PHASE` boundary: consolidate Fisher during sleep, enforce EWC constraint during wake learning cycles.

**Proposed integration point**:
```
SLEEP_PHASE tick N:   MemoryConsolidator.consolidate()
                   +  EWCKernel.update_fisher(recent_weights)
LEARNING tick N+1: STDPOnlineLearner.on_weight_update()
                   +  EWCKernel.apply_constraint(delta)
```

---

### Recommendation

For Phase 5 MVP:
1. **Implement Option A** (kernel protocol) — minimal surface, backward-compatible
2. Ship `STDPKernel` as default (no change to existing design)
3. Add `OjaKernel` as second implementation (simple, tests pluggability)
4. Defer `EWCKernel` to Phase 6 (requires Fisher matrix infra, new issue)

---

### Open Questions

1. **BCM sliding threshold**: θ_M is a running average of postsynaptic activity — where does it live? Blackboard? Learner instance state?
2. **EWC Fisher update cost**: Computing diagonal Fisher requires a backward pass over recent training data. Is this acceptable in a `SLEEP_PHASE` tick?
3. **Kernel hot-swap**: If the operator changes `kernel: stdp → ewc` in config, should the learner reset its weight history or continue from current weights?

Related: Issue #181, Discussion #227, Discussion #231, Issue #233, wiki Phase-5-Online-Learning.md
